File: Day 1/app/src/main/java/com/akshar/todolist/AddTodoFragment.java
package com.akshar.todolist;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageView;

import com.akshar.todolist.model.ToDo;
import com.akshar.todolist.utils.Utils;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;

import java.util.Objects;

public class AddTodoFragment extends BottomSheetDialogFragment {

    private TextInputEditText title, description, date, time;
    private ImageView close;
    private MaterialButton addTodo,saveBtn;
    private CheckBox completed;
    private ToDo toDo;
    public AddTodoFragment() {
    }

    // constructor that accepts a ToDo object
    public AddTodoFragment(ToDo toDo) {
        this.toDo = toDo;
    }



    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.add_todo_sheet, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        title = view.findViewById(R.id.todoTitle);
        description = view.findViewById(R.id.todoDescription);
        date = view.findViewById(R.id.todoDate);
        time = view.findViewById(R.id.todoTime);
        addTodo = view.findViewById(R.id.addTodo);
        close = view.findViewById(R.id.close);
        saveBtn = view.findViewById(R.id.saveTodo);
        completed = view.findViewById(R.id.todoCheckBox);

        if (toDo != null) {
            showDetails();
        } else {
            showAddTodo();
        }
    }

    private void showAddTodo() {
        saveBtn.setVisibility(View.GONE);
        // set current date and time
        date.setText(Utils.getCurrentDate());
        time.setText(Utils.getCurrentTime());

        close.setOnClickListener(v -> dismiss());

        addTodo.setOnClickListener(v -> {
            String todoTitle = Objects.requireNonNull(title.getText()).toString();
            String todoDescription = Objects.requireNonNull(description.getText()).toString();
            String todoDate = Objects.requireNonNull(date.getText()).toString();
            String todoTime = Objects.requireNonNull(time.getText()).toString();

            if (todoTitle.isEmpty()) {
                title.setError("Title cannot be empty");
                return;
            }

            if (todoDescription.isEmpty()) {
                description.setError("Description cannot be empty");
                return;
            }

            if (todoDate.isEmpty()) {
                date.setError("Date cannot be empty");
                return;
            }

            if (todoTime.isEmpty()) {
                time.setError("Time cannot be empty");
                return;
            }

            ToDo toDo = new ToDo(todoTitle, todoDescription, todoDate, todoTime, todoTime, completed.isChecked());
            ((MainActivity) requireActivity()).addTodoItem(toDo);
            dismiss();
        });
    }

    private void showDetails() {
        title.setText(toDo.getTitle());
        description.setText(toDo.getDescription());
        date.setText(toDo.getDate());
        time.setText(toDo.getTime());
        completed.setChecked(toDo.isCompleted());
        disableFields();
        addTodo.setText("Edit");
        saveBtn.setVisibility(View.VISIBLE);

        close.setOnClickListener(v -> dismiss());

        addTodo.setOnClickListener(v -> enableFields());
        saveBtn.setOnClickListener(v -> {
            String todoTitle = Objects.requireNonNull(title.getText()).toString();
            String todoDescription = Objects.requireNonNull(description.getText()).toString();
            String todoDate = Objects.requireNonNull(date.getText()).toString();
            String todoTime = Objects.requireNonNull(time.getText()).toString();

            if (todoTitle.isEmpty()) {
                title.setError("Title cannot be empty");
                return;
            }

            if (todoDescription.isEmpty()) {
                description.setError("Description cannot be empty");
                return;
            }

            if (todoDate.isEmpty()) {
                date.setError("Date cannot be empty");
                return;
            }

            if (todoTime.isEmpty()) {
                time.setError("Time cannot be empty");
                return;
            }

            toDo.setTitle(todoTitle);
            toDo.setDescription(todoDescription);
            toDo.setDate(todoDate);
            toDo.setTime(todoTime);
            toDo.setCompleted(completed.isChecked());

            ((MainActivity) requireActivity()).updateTodoItem(toDo);
            dismiss();
        });


    }

    private void enableFields() {
        title.setEnabled(true);
        description.setEnabled(true);
        date.setEnabled(true);
        time.setEnabled(true);
    }

    private void disableFields() {
        title.setEnabled(false);
        description.setEnabled(false);
        date.setEnabled(false);
        time.setEnabled(false);
    }
}

File: Day 2/app/src/main/java/com/akshar/todoapp/AddTodoFragment.kt
package com.akshar.todoapp

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import com.akshar.todoapp.model.ToDo
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText

class AddTodoFragment : BottomSheetDialogFragment {
    private var title: TextInputEditText? = null
    private var description: TextInputEditText? = null
    private var date: TextInputEditText? = null
    private var time: TextInputEditText? = null
    private var close: ImageView? = null
    private var addTodo: MaterialButton? = null
    private var saveBtn: MaterialButton? = null
    private var completed: CheckBox? = null
    private var toDo: ToDo? = null

    constructor()
    constructor(toDo: ToDo?) {
        this.toDo = toDo
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.add_todo_sheet, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        title = view.findViewById(R.id.todoTitle)
        description = view.findViewById(R.id.todoDescription)
        date = view.findViewById(R.id.todoDate)
        time = view.findViewById(R.id.todoTime)
        addTodo = view.findViewById(R.id.addTodo)
        close = view.findViewById(R.id.close)
        saveBtn = view.findViewById(R.id.saveTodo)
        completed = view.findViewById(R.id.todoCheckBox)
        if (toDo != null) {
            showDetails()
        } else {
            showAddTodo()
        }
    }

    private fun showAddTodo() {
        saveBtn?.visibility = View.GONE
        // set current date and time
        date?.setText(Utils.currentDate)
        time?.setText(Utils.currentTime)
        close?.setOnClickListener { dismiss() }
        addTodo?.setOnClickListener {
            val todoTitle: String = title?.text.toString()
            val todoDescription: String = description?.text.toString()
            val todoDate: String = date?.text.toString()
            val todoTime: String = time?.text.toString()
            if (todoTitle.isEmpty()) {
                title?.error = "Title cannot be empty"
                return@setOnClickListener
            }
            if (todoDescription.isEmpty()) {
                description?.error = "Description cannot be empty"
                return@setOnClickListener
            }
            if (todoDate.isEmpty()) {
                date?.error = "Date cannot be empty"
                return@setOnClickListener
            }
            if (todoTime.isEmpty()) {
                time?.error = "Time cannot be empty"
                return@setOnClickListener
            }
            val toDo = ToDo(
                id = 0, // Assuming 0 as default id
                title = todoTitle,
                description = todoDescription,
                date = todoDate,
                time = todoTime,
                completed = completed?.isChecked ?: false
            )
            (requireActivity() as MainActivity).addTodoItem(toDo)
            dismiss()
        }
    }

    private fun showDetails() {
        title?.setText(toDo?.title)
        description?.setText(toDo?.description)
        date?.setText(toDo?.date)
        time?.setText(toDo?.time)
        completed?.setChecked(toDo?.completed ?: false)
        disableFields()
        addTodo?.text = getString(R.string.edit)
        saveBtn?.visibility = View.VISIBLE
        close?.setOnClickListener { dismiss() }
        addTodo?.setOnClickListener { enableFields() }
        saveBtn?.setOnClickListener {
            val todoTitle: String = title?.text.toString()
            val todoDescription: String = description?.text.toString()
            val todoDate: String = date?.text.toString()
            val todoTime: String = time?.text.toString()
            if (todoTitle.isEmpty()) {
                title?.error = "Title cannot be empty"
                return@setOnClickListener
            }
            if (todoDescription.isEmpty()) {
                description?.error = "Description cannot be empty"
                return@setOnClickListener
            }
            if (todoDate.isEmpty()) {
                date?.error = "Date cannot be empty"
                return@setOnClickListener
            }
            if (todoTime.isEmpty()) {
                time?.error = "Time cannot be empty"
                return@setOnClickListener
            }
            toDo?.title = todoTitle
            toDo?.description = todoDescription
            toDo?.date = todoDate
            toDo?.time = todoTime
            toDo?.completed = completed?.isChecked ?: false
            toDo?.let { it1 -> (requireActivity() as MainActivity).updateTodoItem(it1) }
            dismiss()
        }
    }
    private fun enableFields() {
        title?.isEnabled = true
        description?.isEnabled = true
        date?.isEnabled = true
        time?.isEnabled = true
    }

    private fun disableFields() {
        title?.isEnabled = false
        description?.isEnabled = false
        date?.isEnabled = false
        time?.isEnabled = false
    }
}